src/server/qgsconfigcache.h

```diff
@@ -65,21 +105,155 @@ class SERVER_EXPORT QgsConfigCache : public QObject
 
-  private:
-    QgsConfigCache() SIP_FORCE;
+    /**
+     * Set the invalidation strategy for the cache
+     * The cache takes ownership of the strategy which will be deleted on cleanup
+     * \param strategy the strategy to use
+     * \since QGIS 3.24
+     */
+    void setStrategy( QgsAbstractCacheStrategy *strategy SIP_TRANSFER );
 
-    //! Check for configuration file updates (remove entry from cache if file changes)
-    QFileSystemWatcher mFileSystemWatcher;
+    /**
+     * Return the name of the current strategy
+     * \since QGIS 3.24
+     */
+    QString strategyName() const { return mStrategy->ident(); }
+
+  private:
+    QgsConfigCache( QgsAbstractCacheStrategy *defaultStrategy ) SIP_FORCE;
 
     //! Returns xml document for project file / sld or 0 in case of errors
     QDomDocument *xmlDocument( const QString &filePath );
 
     QCache<QString, QDomDocument> mXmlDocumentCache;
-    QCache<QString, QgsProject> mProjectCache;
+    QCache<QString, std::pair<QDateTime, std::unique_ptr<QgsProject> > > mProjectCache;
+
+    std::unique_ptr<QgsAbstractCacheStrategy> mStrategy;
+
+  private:
+    //! Insert project in cache
+    void cacheProject( const QString &path, QgsProject *project );
 
   private slots:
-    //! Removes changed entry from this cache
+    //! Remove entry
     void removeChangedEntry( const QString &path );
+
+    //! Remove all cached entries
+    void removeChangedEntries();
+
 };
 
+
+/**
+ * \ingroup server
+ * \brief File system cache strategy for server configuration
+ * \since QGIS 3.24
+ */
+class SERVER_EXPORT QgsFileSystemCacheStrategy : public QgsAbstractCacheStrategy
+{
+  public:
+    QgsFileSystemCacheStrategy();
+
+    //! The name of the strategy
+    QString ident() const override { return QStringLiteral( "filesystem" ); };
+
+  protected:
+    //! Attach cache to this strategy
+    void attach( QgsConfigCache *cache ) override;
+
+    /**
+     * Called when an entry is removed from cache.
+     * \param path The path of the project
+     */
+    void entryRemoved( const QString &path ) override;
+
+    //! Called when an entry is inserted
+    void entryInserted( const QString &path ) override;
+
+  private:
+    //! For invalidating files
+    QFileSystemWatcher mFileSystemWatcher;
+};
+
+
+/**
+ * \ingroup server
+ * \brief Periodic system cache strategy for server configuration
+ * \since QGIS 3.24
+ */
+class SERVER_EXPORT QgsPeriodicCacheStrategy: public QgsAbstractCacheStrategy
+{
+  public:
+    QgsPeriodicCacheStrategy( int interval = 3000 );
+
+    //! The name of the strategy
+    QString ident() const override { return QStringLiteral( "periodic" ); };
+
+    /**
+     *src/server/qgsserver.cpp Set the invalidation check interval for PeriodicStrategy
+     *
+     * If \a msec is set to a positive value then the cache will check for invalidation
+     * every \a msec milliseconds
+     *
+     * If \a msec is set to 0 (or negative value), which is the default,
```

Comment:
I'm not convinced about the side-effect of setting the interval, the function doc say that setting it to 0 automatically (de)activates the 'periodic' strategy and activates the 'filesystem' strategy, I think it would be more intuitive to remove this side-effect: setting it to 0 would just effectively deactivate the periodic timer and implicitly deactivate any caching removal strategy (a log message would be probably useful in that case).

Reply:
> That setting it to 0 automatically (de)activates the 'periodic' strategy and activates the 'filesystem' strategy

This is not the case anymore, there is no side-effects between strategies parameters in the current implementation. Setting the strategy is now explicit.

I need to check the documentation.

Reply:
Fixed the documentation.